main/lsp/requests/code_action.cc

```diff
@@ -29,8 +34,234 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
+    auto maybeTree = findMethodTree(rootTree, method.asMethodRef());
+    if (!maybeTree.has_value()) {
+        return nullopt;
+    }
+    auto methodLoc = maybeTree.value()->loc;
+
+    auto maybeSig = sorbet::sig_finder::findSignature(gs, method);
+    if (!maybeSig.has_value()) {
+        return make_pair(nullopt, methodLoc);
+    }
+    core::LocOffsets sigLoc = {maybeSig->sig.beginPos(), maybeSig->body.endPos()};
+
+    return make_pair(sigLoc, methodLoc);
+}
+
+// Turns ruby_function_name__ to RubyFunctionName,
+// so we can use it in a new module name
+string snakeToCamelCase(string_view name) {
+    string res;
+    const auto originalSize = name.size();
+    res.reserve(originalSize);
+    bool shouldCapitalize = true;
+    for (int i = 0; i < originalSize; i++) {
+        if (name.at(i) == '_') {
+            shouldCapitalize = true;
+            continue;
+        }
+        res += shouldCapitalize ? toupper(name[i]) : name[i];
+        shouldCapitalize = false;
+    }
+    return res;
+}
+
+optional<string> getNewModuleName(const core::GlobalState &gs, const core::NameRef name) {
+    const auto moduleName = absl::StrCat(snakeToCamelCase(name.show(gs)), "Module");
+    const auto nameNotExists = [&](auto moduleName) {
+        if (auto nameRef = gs.lookupNameUTF8(moduleName); !nameRef.exists()) {
+            // We're using root here, because newely defined module would be at the top level
+            return !core::Symbols::root().data(gs)->findMember(gs, nameRef).exists();
+        }
+        return false;
+    };
+
+    // if there are no names like that, return the new module name
+    if (nameNotExists(moduleName)) {
+        return moduleName;
+    }
+
+    // otherwise, try to augment it
+    for (int i = 1; i < 42; i++) {
+        const auto newName = absl::StrCat(moduleName, i);
+        if (nameNotExists(newName)) {
+            return newName;
+        }
+    }
+
+    // bail, if we haven't found untaken name in 42 tries
+    return nullopt;
+}
+
+class MethodCallSiteRenamer : public AbstractRenamer {
+public:
+    MethodCallSiteRenamer(const core::GlobalState &gs, const LSPConfiguration &config, const string oldName,
+                          const string newName)
+        : AbstractRenamer(gs, config, oldName, newName) {
+        const vector<string> invalidNames = {"initialize", "call"};
+        for (auto name : invalidNames) {
+            if (oldName == name) {
+                invalid = true;
+                error = fmt::format("The `{}` method cannot be moved to a module.", oldName);
+                return;
+            }
+        }
+        // block any method not starting with /[a-zA-Z0-9_]+/. This blocks operator overloads.
+        if (!isalnum(oldName[0]) && oldName[0] != '_') {
+            error = fmt::format("The `{}` method cannot be moved to a module.", oldName);
+            invalid = true;
+        }
```

Comment:
Do you think it's worth writing tests for these?

I believe you cargo-culted these from `rename.cc` if I'm not mistaken.

Reply:
Actually seems like limitation for the `initialize` and `call` method doesn't make a lot of sense with the current implementation, because the refactoring action is limited to class methods only, and those methods are instance methods. However, I would like to leave this code here, because we might add support for instance methods later. I'll add test we don't generate a code action for those methods.

And testing for the operator overloads is a great idea, I'll do that too.